jetty-util/src/main/java/org/eclipse/jetty/util/resource/ZipFsResource.java

```diff
@@ -0,0 +1,306 @@
+
+    public ZipFsResource(Path path) throws IOException
+    {
+        this(path.toUri(), MR_MODE_RUNTIME);
+    }
+
+    public ZipFsResource(File file) throws IOException
+    {
+        this(file.toURI(), MR_MODE_RUNTIME);
+    }
+
+    public ZipFsResource(URI uri) throws IOException
+    {
+        this(uri, MR_MODE_RUNTIME);
+    }
+
+    public ZipFsResource(URI uri, String multiReleaseMode) throws IOException
+    {
+        zipFsUri = toZipFsUri(uri);
+        URI fileUri = URI.create(uri.getSchemeSpecificPart());
+        zipFsPath = new File(fileUri.getSchemeSpecificPart()).toPath();
+
+        // The FileSystem Environment
+        env = new HashMap<>();
+        env.put("multi-release", multiReleaseMode);
+
+        // The ZipFs FileSystem
+        FileSystem fs;
+        try
+        {
+            fs = FileSystems.getFileSystem(zipFsUri);
+        }
+        catch (FileSystemNotFoundException ignore)
+        {
+            fs = null;
+        }
+
+        if ((fs == null) || !fs.isOpen())
+        {
+            fs = FileSystems.newFileSystem(zipFsUri, env);
+        }
+
+        zipFs = fs;
+
+        // We only care about the 1 root that a zip/jar has
+        Iterable<Path> rootDirs = zipFs.getRootDirectories();
+        rootPath = rootDirs.iterator().next();
+    }
+
+    private final URI toZipFsUri(URI uri)
+    {
+        String asciiUri = uri.toASCIIString();
+
+        if (asciiUri.equalsIgnoreCase("jar:file:") &&
+                asciiUri.endsWith(".jar"))
+        {
+            // we have what we need already.
+            return uri;
+        }
+
+        if ("file".equalsIgnoreCase(uri.getScheme()))
+        {
+            if (asciiUri.endsWith(".jar"))
+            {
+                return URI.create("jar:" + asciiUri);
+            }
+        }
+
+        throw new UnsupportedOperationException(ZipFsResource.class.getSimpleName() + " only supports [jar:file:/] or [file:/] URIs that end with [.jar] - " + uri);
+    }
+
+    @Override
+    public boolean isContainedIn(Resource resource) throws MalformedURLException
+    {
+        if (resource instanceof PathResource)
+        {
+            PathResource pathResource = (PathResource) resource;
+            Path internalPath = zipFs.getPath(resource.getName());
+            try
+            {
+                return Files.isSameFile(pathResource.getPath(), internalPath);
+            }
+            catch (IOException ignore)
+            {
+                LOG.ignore(ignore);
+                return false;
+            }
+        }
+        // all other resource types are not supported here
+        return false;
+    }
+
+    @Override
+    public void close()
+    {
+        try
+        {
+            zipFs.close();
+        }
+        catch (IOException ignore)
+        {
+            LOG.ignore(ignore);
+        }
+    }
+
+    @Override
+    public boolean exists()
+    {
+        // If we passed the constructor, this is known to exist
```

Comment:
really? jar:file:///some/path/file.zip!/other/path/fileThatDoesNotExist.html should be able to be created by addPath and it might not exist!

Reply:
That would be a `java.nio.file.Path` using `PathResource` and not a `ZipFsResource`